mvcc: allow clients to assign watcher IDs

connor4312 opened the pull request:
This allows for watchers to be created concurrently without needing complex and latency-adding queuing on the client.

I purposely didn't update clientv3 yet, in case consumers import an updated clientv3 but don't run the latest etcd release. This would cause subtly breaking behavior for them. Tested this end to end locally and it seems to work fine.

Resolves https://github.com/coreos/etcd/issues/7036

connor4312 commented:
The built failure seems unrelated to the changes in this PR, as I'm seeing it on previous PRs like [this one](https://github.com/coreos/etcd/pull/8656) which just updates the readme

xiang90 commented on mvcc/watcher.go:
```diff
@@ -23,7 +23,8 @@ import (
 var (
-	ErrWatcherNotExist = errors.New("mvcc: watcher does not exist")
+	ErrWatcherNotExist    = errors.New("mvcc: watcher does not exist")
+	ErrWatcherDuplicateID = errors.New("mvcc: duplicate ID on the WatchStream")
```
duplicate watch ID provided on the WatchStream

xiang90 commented on mvcc/watcher.go:
```diff
@@ -33,15 +34,17 @@ type FilterFunc func(e mvccpb.Event) bool
 	// in events that are sent to the created watcher through stream channel.
 	//
-	Watch(key, end []byte, startRev int64, fcs ...FilterFunc) WatchID
+	Watch(id WatchID, key, end []byte, startRev int64, fcs ...FilterFunc) (WatchID, error)
```
we need to define the nonWatchID as a const.

xiang90 commented on mvcc/watcher.go:
```diff
@@ -33,15 +34,17 @@ type FilterFunc func(e mvccpb.Event) bool
 	// Watch creates a watcher. The watcher watches the events happening or
-	// happened on the given key or range [key, end) from the given startRev.
+	// happened on the given key or range [key, end) from the given startRev. If
+	// the ID is provided and non-zero, it will be used, otherwise a new ID
```
it seems that -1 is the value, not non-zero.

connor4312 replied:
So, it actually is zero, because it doesn't look like the protobuf is set to have optional types (so, if the watcher ID is not present in the WatchRequest, it's zero). There is some kind of weird casing around 0 and -1 right now, since I didn't change the -1 return the function had previously.

Instead, I want to normalize around 0 and get rid of the -1 return in favor of a sentinel error.

xiang90 commented on mvcc/watcher_test.go:
```diff
@@ -79,6 +79,39 @@ func TestWatcherWatchID(t *testing.T) {
+	// - Try to duplicate it, get an error
+	// - Make sure the auto-assignment skips over things we manually assigned
+
+	id, err := w.Watch(1, []byte("foo"), nil, 0)
```
can you convert this to a table driven test?

xiang90 commented:
lgtm in general.

connor4312 commented:
Thanks for the review! I'll get it adjusted tonight

xiang90 commented:
@connor4312 kindly ping.

connor4312 commented:
Hi, I've updated my PR. A few changes:

 - Remove the magic `-1`. Made empty/"bad" ranges return a sentinel error instead of signalling with -1.
 - The zero ID is the only magic ID, for which we'll assign an ID automatically (no way to tell from the protobuf the difference between a 0 and a not-present field)
 - Moved the bulk of the logic in recvLoop out to its own recvWatchRequest method which simplifies the control flow a good bit
 - Moved the mentioned test to a table driven test

xiang90 commented on mvcc/watcher.go:
```diff
@@ -22,8 +22,14 @@ import (
 
+// DefaultWatchID is the watcher ID passed in WatchStream.Watch when no
+// user-provided ID is available. If pass, an ID will automatically be assigned.
+const DefaultWatchID WatchID = 0
```
AutoWatchID? it is more descriptive. when auto watch id is set, etcd automatically assign watch id.

xiang90 commented:
@connor4312 looks good to me in general. can you rebase with the current master so test can pass? thanks.

connor4312 commented:
Done and done :)

xiang90 commented:
@connor4312 Can you rebase with the current master to let CI pass? Thanks!

gyuho commented:
@connor4312 Sorry for delay! Our CIs have been broken for so long, and finally it's fixed now.
I cherry-picked this patch to resolve conflicts and also to include this in our upcoming 3.3 release.

https://github.com/coreos/etcd/pull/9015
